common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs

```diff
@@ -67,14 +80,31 @@ public class NewtonSoftJsonSerializer : IJsonSerializerSettingsProvider,
         }
 
         /// <inheritdoc/>
-        public object Deserialize(ReadOnlyMemory<byte> buffer, Type type) {
+        public object Deserialize(ReadOnlyMemory<byte> buffer, Type type, TextReader schemaReader = null) {
             try {
+
                 // TODO move to .net 3 to use readonly span as stream source
-                var jsonSerializer = JsonSerializer.CreateDefault(Settings);
-                using (var stream = new MemoryStream(buffer.ToArray()))
-                using (var reader = new StreamReader(stream, ContentEncoding)) {
-                    return jsonSerializer.Deserialize(reader, type);
+                var bufferArray = buffer.ToArray();
+
+                // Validate json if schema is provided.
+                if (schemaReader != null) {
+                    var validationResults = _jsonSchemaValidator.Validate(bufferArray, schemaReader);
+
+                    // If any item in validation results collection is marked as invalid, consider it as json schema validation failure.
+                    if (validationResults.Any(r => r.IsValid == false)) {
+                        var validationResultsMessage =
+                            string.Join("; ",
+                                validationResults.Select(r =>
+                                    $"Validation failed with error: {r.Message} at schema path: {r.SchemaLocation}, and configuration file location {r.InstanceLocation}"));
+
+                        throw new JsonSerializationException(validationResultsMessage);
```

Comment:
`JsonSerializationException` object created here will be caught on line `109` bellow and transformed into `SerializerException` object. So I would suggest to throw `SerializerException` here instead.

Reply:
Thanks, it was done this way to avoid the breaking change i.e. existing deserialization code would throw `JsonSerializationException` which is wrapped by `SerializerException`. By keeping the same wrapping with inner exception being `JsonSerializationException` we maintain the callers treatment of exceptions from this method.

Reply:
OK, got you. Then let's keep it as it is.